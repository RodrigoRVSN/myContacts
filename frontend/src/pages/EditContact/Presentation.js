import React from 'react';
import PropTypes from 'prop-types';
import { ContactForm } from '../../components/ContactForm';
import { Loader } from '../../components/Loader';
import { PageHeader } from '../../components/PageHeader';

export default function Presentation({
  isLoading, contactFormRef, contactName, handleSubmit,
}) {
  return (
    <>
      <Loader isLoading={isLoading} />

      <PageHeader title={isLoading ? 'Carregando...' : `Editar ${contactName}`} />

      <ContactForm buttonLabel="Salvar" ref={contactFormRef} onSubmit={handleSubmit} />
    </>
  );
}

Presentation.propTypes = {
  isLoading: PropTypes.bool.isRequired,
  contactFormRef: PropTypes.object.isRequired,
  contactName: PropTypes.string.isRequired,
  handleSubmit: PropTypes.func.isRequired,
};
