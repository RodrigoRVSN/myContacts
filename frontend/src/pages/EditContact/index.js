import React from 'react';
import { ContactForm } from '../../components/ContactForm';
import { PageHeader } from '../../components/PageHeader';

export function EditContact() {
  return (
    <>
      <PageHeader title="Editar Contato" />
      <ContactForm buttonLabel="Salvar" />
    </>
  );
}
