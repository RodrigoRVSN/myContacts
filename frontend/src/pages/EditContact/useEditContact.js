import { useState, useEffect, useRef } from 'react';
import { useHistory, useParams } from 'react-router-dom';
import { useSafeAsyncAction } from '../../hooks/useSafeAsyncAction';
import ContactsService from '../../services/ContactsService';
import { toast } from '../../utils/toast';

export const useEditContact = () => {
  const [isLoading, setIsLoading] = useState(true);
  const [contactName, setContactName] = useState('');
  const { id } = useParams();
  const history = useHistory();
  const contactForm = useRef(null);
  const safeAsyncAction = useSafeAsyncAction();

  useEffect(() => {
    const controller = new AbortController();

    const loadContacts = async () => {
      try {
        const contact = await ContactsService.getContactById(id, controller.signal);

        safeAsyncAction(() => {
          contactForm.current.setFieldValues(contact);
          setContactName(contact.name);
          setIsLoading(false);
        });
      } catch (error) {
        if (!(error instanceof DOMException && error.name === 'AbortError')) {
          safeAsyncAction(() => {
            history.push('/');
            toast({ type: 'danger', text: 'Contato nao existe!' });
          });
        }
      }
    };

    loadContacts();

    return () => {
      controller.abort();
    };
  }, [history, id, safeAsyncAction]);

  async function handleSubmit(contact) {
    try {
      const { name } = await ContactsService.updateContact(id, contact);
      setContactName(name);
      toast({ type: 'success', text: 'Contato editado com sucesso!', duration: 10000 });
    } catch (error) {
      toast({ type: 'danger', text: 'Houve um erro!' });
    }
  }

  return {
    isLoading, contactName, contactForm, handleSubmit,
  };
};
