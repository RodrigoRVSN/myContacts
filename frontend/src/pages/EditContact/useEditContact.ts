import { useState, useEffect, useRef } from 'react';
import { useNavigate, useParams } from 'react-router-dom';
import { useSafeAsyncAction } from '../../hooks/useSafeAsyncAction';
import ContactsService from '../../services/ContactsService';
import { DomainContact, IContact } from '../../types/IContact';
import { toast } from '../../utils/toast';

export const useEditContact = () => {
  const [isLoading, setIsLoading] = useState(true);
  const [contactName, setContactName] = useState('');
  const { id } = useParams<{ id: string }>();

  const navigate = useNavigate();
  const contactForm = useRef<{setFieldValues: (contact: IContact) => void}>(null);
  const safeAsyncAction = useSafeAsyncAction();

  useEffect(() => {
    const controller = new AbortController();

    const loadContacts = async () => {
      try {
        const contact = await ContactsService.getContactById(String(id), controller.signal);

        safeAsyncAction(() => {
          contactForm.current!.setFieldValues(contact);
          setContactName(contact.name);
          setIsLoading(false);
        });
      } catch (error) {
        if (!(error instanceof DOMException && error.name === 'AbortError')) {
          safeAsyncAction(() => {
            navigate('/');
            toast({ type: 'danger', text: 'Contato nao existe!' });
          });
        }
      }
    };

    loadContacts();

    return () => {
      controller.abort();
    };
  }, [history, id, safeAsyncAction]);

  async function handleSubmit(contact: DomainContact) {
    try {
      const { name } = await ContactsService.updateContact(String(id), contact);
      setContactName(name);
      toast({ type: 'success', text: 'Contato editado com sucesso!', duration: 10000 });
    } catch (error) {
      toast({ type: 'danger', text: 'Houve um erro!' });
    }
  }

  return {
    isLoading, contactName, contactForm, handleSubmit,
  };
};
