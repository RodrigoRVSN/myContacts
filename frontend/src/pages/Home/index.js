/* eslint-disable no-nested-ternary */
import React, {
  useCallback, useEffect, useMemo, useState,
} from 'react';
import { Link } from 'react-router-dom';
import {
  Container,
  Header,
  Card,
  InputSearchContainer,
  ListHeader,
  ErrorContainer,
  NoContactsContainer,
  NotFoundFilteredContacts,
} from './styles';

import arrow from '../../assets/images/icons/arrow.svg';
import edit from '../../assets/images/icons/edit.svg';
import trash from '../../assets/images/icons/trash.svg';
import sad from '../../assets/images/sad.svg';
import emptyBox from '../../assets/images/empty-box.svg';
import magnifierQuestion from '../../assets/images/magnifier-question.svg';

import { Loader } from '../../components/Loader';
import ContactsService from '../../services/ContactsService';
import Button from '../../components/Button';

export function Home() {
  const [contacts, setContacts] = useState([]);
  const [orderBy, setOrderBy] = useState('asc');
  const [searchTerm, setSearchTerm] = useState('');
  const [loading, setLoading] = useState(true);
  const [hasError, setError] = useState(false);

  const filteredContacts = useMemo(
    () =>
      contacts.filter((contact) =>
        contact.name.toLowerCase().includes(searchTerm.toLowerCase())),
    [contacts, searchTerm],
  );

  const loadContacts = useCallback(async () => {
    try {
      setLoading(true);
      const contactsList = await ContactsService.listContacts(orderBy);

      setError(false);
      setContacts(contactsList);
    } catch (error) {
      setError(true);
    } finally {
      setLoading(false);
    }
  }, [orderBy]);

  useEffect(() => {
    loadContacts();
  }, [loadContacts]);

  function handleToggleOrderBy() {
    setOrderBy((prevState) => (prevState === 'asc' ? 'desc' : 'asc'));
  }

  function handleChangeSearchTerm(ev) {
    setSearchTerm(ev.target.value);
  }

  function handleTryAgain() {
    loadContacts();
  }

  return (
    <>
      <Container>
        <Loader isLoading={loading} />

        {contacts.length > 0 && (
          <InputSearchContainer>
            <input
              value={searchTerm}
              onChange={(ev) => handleChangeSearchTerm(ev)}
              type="text"
              placeholder="Procure pelo contato"
            />
          </InputSearchContainer>
        )}

        <Header justifyContent={
          hasError
            ? 'flex-end'
            : (contacts.length > 0
              ? 'space-between'
              : 'center')
        }
        >
          {(!hasError && contacts.length > 0) && (
            <strong>
              {filteredContacts.length > 0
                ? `${filteredContacts.length} ${filteredContacts.length > 1 ? ' contatos' : ' contato'
                }`
                : 'Nenhum contato!'}
            </strong>
          )}
          <Link to="/new">Novo Contato</Link>
        </Header>

        {hasError && (
          <ErrorContainer>
            <img src={sad} alt="Sad" />
            <div className="error_container">
              <strong>Ocorreu um erro ao obter os seus contatos!</strong>
              <Button onClick={handleTryAgain}>Tentar novamente</Button>
            </div>
          </ErrorContainer>
        )}

        {!hasError && (
          <>
            {contacts.length < 1 && (
              <NoContactsContainer>
                <img src={emptyBox} alt="Sem contatos" />
                <p>
                  Você ainda não tem nenhum contato cadastrado!
                  Clique no botão <strong>”Novo contato”</strong> à cima
                  para cadastrar o seu primeiro!
                </p>
              </NoContactsContainer>
            )}

            {filteredContacts.length > 0 && (
              <ListHeader orderBy={orderBy}>
                <button onClick={handleToggleOrderBy} type="button">
                  <span>Nome</span>
                  <img src={arrow} alt="Arrow" />
                </button>
              </ListHeader>
            )}

            {(contacts.length > 0 && filteredContacts.length < 1)
              && (
                <NotFoundFilteredContacts>
                  <img src={magnifierQuestion} alt="magnifierQuestion" />
                  <span>
                    Nenhum resultado foi encontrado para <b>{searchTerm}</b>.
                  </span>
                </NotFoundFilteredContacts>
              )}

            {filteredContacts.map((contact) => (
              <Card key={contact.id}>
                <div className="info">
                  <div className="contact-name">
                    <strong>{contact.name}</strong>
                    {contact.category_name && (
                      <small>{contact.category_name}</small>
                    )}
                  </div>
                  <span>{contact.email}</span>
                  <span>{contact.phone}</span>
                </div>
                <div className="actions">
                  <Link to={`/edit/${contact.id}`}>
                    <img src={edit} alt="edit" />
                  </Link>
                  <button type="button">
                    <img src={trash} alt="trash" />
                  </button>
                </div>
              </Card>
            ))}
          </>
        )}

      </Container>
    </>
  );
}
