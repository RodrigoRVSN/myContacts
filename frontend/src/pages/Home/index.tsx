import React from 'react';
import {
  Container,
} from './styles';

import { Loader } from '../../components/Loader';
import { Modal } from '../../components/Modal';
import { useHome } from './useHome';
import { InputSearch } from './components/InputSearch';
import { Header } from './components/Header';
import { ErrorStatus } from './components/ErrorStatus';
import { EmptyList } from './components/EmptyList';
import { SearchNotFound } from './components/SearchNotFound';
import { ContactsList } from './components/ContactsList';

export function Home() {
  const {
    isLoading,
    isDeleteModalVisible,
    contactBeingDeleted,
    handleToggleDeleteModal,
    handleConfirmDeleteContact,
    isLoadingDelete,
    contacts,
    searchTerm,
    handleChangeSearchTerm,
    hasError,
    filteredContacts,
    handleTryAgain,
    orderBy,
    handleToggleOrderBy,
    handleDeleteContact,
  } = useHome();

  const hasContacts = contacts.length > 0;
  const isListEmpty = !hasError && (!isLoading && !hasContacts);
  const isSearchEmpty = !hasError && hasContacts && filteredContacts.length < 1;

  return (
    <Container>
      <Loader isLoading={isLoading} />

      {hasContacts && (
        <InputSearch
          value={searchTerm}
          onChange={handleChangeSearchTerm}
        />
      )}

      <Header
        hasError={hasError}
        qtyContacts={contacts.length}
        qtyFilteredContacts={filteredContacts.length}
      />

      {hasError && <ErrorStatus onTryAgain={handleTryAgain} />}
      {isListEmpty && <EmptyList />}
      {isSearchEmpty && <SearchNotFound searchTerm={searchTerm} />}

      {hasContacts && (
      <>
        <ContactsList
          filteredContacts={filteredContacts}
          onDeleteContact={handleDeleteContact}
          orderBy={orderBy}
          onToggleOrderBy={handleToggleOrderBy}
        />

        <Modal
          danger
          isVisible={isDeleteModalVisible}
          title={`Tem certeza que deseja remover o contato "${contactBeingDeleted?.name}"?`}
          cancelLabel="Cancelar"
          confirmLabel="Confirmar"
          onCancel={handleToggleDeleteModal}
          onConfirm={handleConfirmDeleteContact}
          isLoading={isLoadingDelete}
        >
          <p>Esta ação não poderá ser desfeita!</p>
        </Modal>
      </>
      )}
    </Container>
  );
}
