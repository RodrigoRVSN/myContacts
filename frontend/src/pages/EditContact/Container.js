import React, { useState, useEffect, useRef } from 'react';
import { useHistory, useParams } from 'react-router-dom';
import { useSafeAsyncAction } from '../../hooks/useSafeAsyncAction';
import ContactsService from '../../services/ContactsService';
import { toast } from '../../utils/toast';
import Presentation from './Presentation';

export default function EditContact() {
  const [isLoading, setIsLoading] = useState(true);
  const [contactName, setContactName] = useState('');
  const { id } = useParams();
  const history = useHistory();
  const contactFormRef = useRef(null);
  const safeAsyncAction = useSafeAsyncAction();

  useEffect(() => {
    const loadContacts = async () => {
      try {
        const contact = await ContactsService.getContactById(id);

        safeAsyncAction(() => {
          contactFormRef.current.setFieldValues(contact);
          setContactName(contact.name);
          setIsLoading(false);
        });
      } catch {
        safeAsyncAction(() => {
          history.push('/');
          toast({ type: 'danger', text: 'Contato nao existe!' });
        });
      }
    };

    loadContacts();
  }, [history, id, safeAsyncAction]);

  async function handleSubmit(contact) {
    try {
      const { name } = await ContactsService.updateContact(id, contact);
      setContactName(name);
      toast({ type: 'success', text: 'Contato editado com sucesso!', duration: 10000 });
    } catch (error) {
      toast({ type: 'danger', text: 'Houve um erro!' });
    }
  }

  return (
    <>
      <Presentation
        isLoading={isLoading}
        contactFormRef={contactFormRef}
        contactName={contactName}
        handleSubmit={handleSubmit}
      />
    </>
  );
}
