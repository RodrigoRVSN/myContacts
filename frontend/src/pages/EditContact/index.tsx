import React from 'react';
import { ContactForm } from '../../components/ContactForm';
import { Loader } from '../../components/Loader';
import { PageHeader } from '../../components/PageHeader';
import { useEditContact } from './useEditContact';

export default function EditContact() {
  const {
    isLoading, contactName, contactForm, handleSubmit,
  } = useEditContact();

  return (
    <>
      <Loader isLoading={isLoading} />

      <PageHeader title={isLoading ? 'Carregando...' : `Editar ${contactName}`} />

      <ContactForm buttonLabel="Salvar" ref={contactForm} onSubmit={handleSubmit} />
    </>
  );
}
