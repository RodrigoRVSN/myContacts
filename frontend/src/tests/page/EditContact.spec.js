import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import { render } from '@testing-library/react';
import { EditContact } from '../../pages/EditContact';

describe('<EditContact />', () => {
  it('should render edit contacts page', () => {
    const { container } = render(
      <Router>
        <EditContact />
      </Router>,
    );
    expect(container).toMatchSnapshot();
  });
});
