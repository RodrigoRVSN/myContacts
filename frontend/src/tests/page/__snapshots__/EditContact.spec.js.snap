// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditContact /> should render edit contacts page 1`] = `
<div>
  <header
    class="sc-gKAaRy fGWbG"
  >
    <a
      href="/"
    >
      <img
        alt="arrow back"
        src="arrow.svg"
      />
      <span>
        Voltar
      </span>
    </a>
    <h1>
      Editar Contato
    </h1>
  </header>
  <form
    class="sc-eCApnc hurtkR"
    novalidate=""
  >
    <div
      class="sc-gtsrHT jaRrUn"
    >
      <input
        class="sc-dlnjwi eBKfJg"
        placeholder="Nome *"
        value=""
      />
    </div>
    <div
      class="sc-gtsrHT jaRrUn"
    >
      <input
        class="sc-dlnjwi eBKfJg"
        placeholder="E-mail"
        type="email"
        value=""
      />
    </div>
    <div
      class="sc-gtsrHT jaRrUn"
    >
      <input
        class="sc-dlnjwi eBKfJg"
        maxlength="15"
        placeholder="Telefone"
        type="text"
        value=""
      />
    </div>
    <div
      class="sc-gtsrHT jaRrUn"
    >
      <select
        class="sc-hKFxyN bgquXc"
      >
        <option
          value=""
        >
          Categoria
        </option>
        <option
          value="discord"
        >
          Discord
        </option>
        <option
          value="instagram"
        >
          Instagram
        </option>
      </select>
    </div>
    <div
      class="sc-jSFjdj eIgFkL"
    >
      <button
        class="sc-bdnxRM ehnIfz"
        disabled=""
        type="submit"
      >
        Salvar
      </button>
    </div>
  </form>
</div>
`;
